// aws-sdk-js-codemod/dist/transforms/v2-to-v3/modules/importEqualsModule/addNamedModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addNamedModule = void 0;
const getImportEqualsDeclarationType_1 = require("../getImportEqualsDeclarationType");
const addDefaultModule_1 = require("./addDefaultModule");
const getDefaultName_1 = require("./getDefaultName");
const addNamedModule = (j, source, options) => {
    const { importedName, localName = importedName, packageName } = options;
    const defaultLocalName = (0, getDefaultName_1.getDefaultName)(packageName);
    const existingDeclaration = source.find(j.TSImportEqualsDeclaration, {
        type: "TSImportEqualsDeclaration",
        id: {
            type: "Identifier",
            name: localName,
        },
        moduleReference: {
            type: "TSQualifiedName",
            left: {
                type: "Identifier",
                name: defaultLocalName,
            },
            right: {
                type: "Identifier",
                name: importedName,
            },
        },
    });
    if (existingDeclaration.size()) {
        return;
    }
    const defaultDeclaration = (0, getImportEqualsDeclarationType_1.getImportEqualsDeclarationType)(packageName);
    if (source.find(j.TSImportEqualsDeclaration, defaultDeclaration).size() === 0) {
        (0, addDefaultModule_1.addDefaultModule)(j, source, packageName);
    }
    const importEqualsDeclaration = j.tsImportEqualsDeclaration(j.identifier(localName), j.tsQualifiedName(j.identifier(defaultLocalName), j.identifier(importedName)));
    const v3ClientImportEquals = source
        .find(j.TSImportEqualsDeclaration, defaultDeclaration)
        .filter((importEqualsDeclaration) => importEqualsDeclaration.value.id.name === defaultLocalName);
    // Insert import equals after the package import equals.
    if (v3ClientImportEquals.size() > 0) {
        v3ClientImportEquals.at(0).insertAfter(importEqualsDeclaration);
        return;
    }
    // Unreachable code, throw error
    throw new Error("The named import equals can't exist on it's own.\n" +
        "Please report your use case on https://github.com/awslabs/aws-sdk-js-codemod");
};
exports.addNamedModule = addNamedModule;
